ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/TypeHelper.java

```diff
@@ -0,0 +1,201 @@
+ */
+
+package com.spectralogic.ds3autogen.c.helpers;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.spectralogic.ds3autogen.api.models.Ds3Element;
+import com.spectralogic.ds3autogen.api.models.Ds3EnumConstant;
+import com.spectralogic.ds3autogen.c.converters.ElementConverter;
+import com.spectralogic.ds3autogen.c.models.Element;
+import com.spectralogic.ds3autogen.c.models.Type;
+import com.spectralogic.ds3autogen.utils.Helper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.ParseException;
+import java.util.stream.Collectors;
+
+import static com.spectralogic.ds3autogen.c.helpers.CHelper.indent;
+import static com.spectralogic.ds3autogen.utils.ConverterUtil.isEmpty;
+
+public class TypeHelper {
+    private static final Logger LOG = LoggerFactory.getLogger(TypeHelper.class);
+    private TypeHelper() {}
+
+    private final static TypeHelper typeHelper = new TypeHelper();
+
+    public static TypeHelper getInstance() {
+        return typeHelper;
+    }
+
+    public static String getNameUnderscores(final String name) {
+        return Helper.camelToUnderscore(Helper.removeTrailingRequestHandler(Helper.unqualifiedName(name)));
+    }
+
+    public static String getDs3Type(final String name) {
+        return "ds3_" + getNameUnderscores(name);
+    }
+
+    public static String getResponseTypeName(final String name) {
+        return getDs3Type(name) + "_response";
+    }
+
+    public static String getParserFunctionName(final String name) {
+        return "_parse_" + getResponseTypeName(name);
+    }
+
+    public static String getFreeFunctionName(final String name) {
+        return getResponseTypeName(name) + "_free";
+    }
+
+    public static String getEnumValues(final ImmutableList<Ds3EnumConstant> enumConstants) {
+        if (isEmpty(enumConstants)) {
+            return "";
+        }
+        return enumConstants
+                .stream()
+                .map(i -> indent(1) + i.getName())
+                .collect(Collectors.joining(",\n"));
+    }
+
+    public static String getTypeElementsList(final ImmutableList<Element> elements) throws ParseException {
+        final StringBuilder outputBuilder = new StringBuilder();
+
+        for (final Element element : elements) {
+            outputBuilder.append(indent(1)).
+                    append(ElementHelper.getDs3Type(element)).
+                    append(" ").
+                    append(ElementHelper.getNameUnderscores(element.getName())).
+                    append(";").
+                    append(System.lineSeparator());
+        }
+
+        return outputBuilder.toString();
+    }
+
+    public static String generateMatcher(final ImmutableList<Ds3EnumConstant> enumConstants) {
+        final StringBuilder outputBuilder = new StringBuilder();
+        final int numConstants = enumConstants.size();
+
+        if (numConstants > 0) {
```

Comment:
Instead of putting this large block inside the if statement, it would be cleaner to reverse it, and have:
if (numConstants <= 0) { return "": }